homeassistant/components/onvif/event.py

```diff
@@ -169,6 +169,8 @@ async def async_pull_messages(self, _now: dt = None) -> None:
                 )
 
                 # Renew subscription if less than two hours is left
+                if not isinstance(response.TerminationTime, dt.datetime):
+                    raise Fault(message="Invalid TerminationTime '%s'" % response.TerminationTime)
```

Comment:
`Fault` is not an exception that should be raised from within Home Assistant, as it is an exception that belongs to the `zeep` module.

Reply:
The following exceptions are catch below in line 181:
SUBSCRIPTION_ERRORS = (Fault, asyncio.TimeoutError, TransportError)
I think Fault is a best choose from this list because the error source is the response generated by zeep. 
I can replace `raise Fault` with `raise TransportError` or add a new type of exception but I don't see any reason to do that.

Reply:
Please compare logs before:
```
2021-07-22 21:32:43 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
File "/usr/src/homeassistant/components/onvif/event.py", line 171, in async_pull_messages
dt_util.as_utc(response.TerminationTime) - dt_util.utcnow()
File "/usr/src/homeassistant/homeassistant/util/dt.py", line 74, in as_utc
if dattim.tzinfo == UTC:
AttributeError: 'NoneType' object has no attribute 'tzinfo'
```
and after this fix:
```
2021-07-23 21:49:05 DEBUG (MainThread) [onvif] Restarted ONVIF PullPoint subscription for '00:12:41:ED:50:80'
2021-07-23 23:05:55 WARNING (MainThread) [onvif] Failed to fetch ONVIF PullPoint subscription messages for '00:12:41:ED:50:80': Invalid TerminationTime 'None'
2021-07-23 23:05:56 DEBUG (MainThread) [onvif] Restarted ONVIF PullPoint subscription for '00:12:41:ED:50:80'
```

Reply:
> The following exceptions are catch below in line 181:
> SUBSCRIPTION_ERRORS = (Fault, asyncio.TimeoutError, TransportError)
> I think Fault is a best choose from this list because the error source is the response generated by zeep.
> I can replace `raise Fault` with `raise TransportError` or add a new type of exception but I don't see any reason to do that.

You can raise a standard exception, or create a new one for this specific TZ error and add it to the Tuple of exceptions that we catch in `SUBSCRIPTION_ERRORS` instead of reusing `Fault`. As @frenck mentioned, that means a specific thing as related to the `zeep` library.

Reply:
There is no a `TZ error`. The object simply is `None` but it is not handled inside dt_util.
Adding an exception for `TZ error` is a wrong way. 
Adding a standard exception is a wrong way too. SUBSCRIPTION_ERRORS is used in many places and adding a new exception to it, especially a standard exception (TypeError, ValueError), may change the current behavior unexpectedly and requires adding new tests. 
Look at logs with added AttributeError:
```
2021-07-23 23:05:55 WARNING (MainThread) [onvif] Failed to fetch ONVIF PullPoint subscription messages for '00:12:41:ED:50:80': 'NoneType' object has no attribute 'tzinfo'
2021-07-23 23:05:56 DEBUG (MainThread) [onvif] Restarted ONVIF PullPoint subscription for '00:12:41:ED:50:80'
```
The error message is confusing and uninformative in this case.
**I agree that it is a bad idea to raise 3rd party exceptions but it is not this case!**
The current code already catches the `Fault` exception anyway. If your standards don't allow to raise `Fault` inside that try block, something is wrong in standards.
I will not make the code worse by adding unnecessary complications and making it even less stable.
I will keep this patch here https://github.com/slydiman/ha_custom_onvif
Thanks for review.

Reply:
Ok, we'll close here for now then. Thanks for your contribution.

Reply:
I am sorry you feel that way. To hopefully better explain, the `Fault` exception in zeep is used when a `<SoapFault>` XML object is returned from the server. In SOAP, a Fault means a very specific thing [as described in the specification](https://www.w3.org/TR/2000/NOTE-SOAP-20000508/#_Toc478383507). It is an error returned directly from the web service, which is why it makes no sense for a Home Assistant integration to throw it.

Exception handling should be explicit. I'm sorry if I didn't fully understand the initial problem, but that point still stands. Create/throw/catch an exception for that particular problem, perhaps `InvalidTimestamp` is more descriptive for the problem at hand? This way, if somebody is debugging a problem, they won't see a `Fault` exception and automatically assume the problem exists on the end of the web service.

Reply:
Missing `TerminationTime` in the response means that something is wrong in web service communication and the logic/behavior is the same as for `Fault`. Raise Fault is a workaround in this case. My patch is simple as possible and safe. It does not require any new tests. Changing SUBSCRIPTION_ERRORS is not safe without tests because it is used in many places. I don't like the idea to extend SUBSCRIPTION_ERRORS with a special exception in one specific place. There is no any reason to add InvalidTimestamp because nobody will see that exception in logs. The error message is `Failed to fetch ONVIF PullPoint subscription messages` + detailed message from the exception, which is `Invalid TerminationTime ...`. BTW, InvalidTimestamp seems more common and it must be added somewhere in dt_util, not in onvif.